prototypes/coin-selection/CoinSelection.hs

```diff
@@ -0,0 +1,278 @@
+    -- wallet's UTxO.
+    | Unbalanced
+    -- ^ Value is added to the first change output until it reaches the maximum
+    -- size. Then the next change output will receive value.
+    deriving (Show, Read, Eq, Enum, Generic)
+
+-- | Placeholder for the type of possible failure cases in coin selection.
+newtype SelectionError = SelectionError String -- ^ Something went wrong.
+    deriving (Show, Read, Eq, Generic)
+
+-- $naming
+-- We call the coin selection function 'xyzzy' in lieu of a better name.
+--
+-- Possible names could be @balanceTransaction@, @selectCoins@, or
+-- something like that.
+
+-- | The top-level coin selection function.
+--
+-- Given an initial partially funded transaction body, it will produce a
+-- complete, funded, and balanced transaction body from the provided 'UTxO'.
+--
+-- If successful, it returns one or more transactions, otherwise the reason for
+-- failure is provided.
+--
+-- __How to use__
+--
+-- The main interface of 'xyzzy' is its parameter @tx@ the partial
+-- transaction. The caller specifies the outputs of @tx@, and 'xyzzy' calculates
+-- a balanced 'TxBody' from that.
+--
+-- [Payment]: Add a 'TxOut' to the initial @tx@ with the recipient
+--   'Address' and 'Value'.
+-- [Change Output]: Add a 'TxOut' to the initial @tx@ such that the 'Address'
+--   satisfies 'isOurAddress'. Its 'Value' would usually be zero.
+--   If the maximum output size is reached, another 'TxOut' with the same
+--   address will be added to the transaction.
+-- [No new change addresses]: If there is no transaction output that satisfies
+--   v'IsOurAddress', then change will be directed back to the addresses
+--   belonging to the wallet which funded the transaction.
+-- [Join stake pool]: This transaction will involve a delegation certificate,
+--   and possibly also a stake key registration certificate. These cost money.
+--   However 'xyzzy', is not necessarily Cardano-specific or even POS-specific.
+--   Therefore, the 'Address' type for @tx@ should include a constructor for
+--   certificates. The implementation of 'appendTxOutput' should put these
+--   "certificate" address outputs in the correct place within the concrete
+--   'TxBody' type.
+-- [Withdraw from reward account]: Reward account withdrawals add value to the
+--   transaction. The 'TxIn' type for @tx@ should include a constructor for
+--   withdrawals. Put this 'TxIn' either in the 'UTxO' list or in the initial
+--   @tx@.
+-- [Coalesce UTxO]: Add a transaction output with an address belonging to the
+--   wallet and pass the v'Sweep' option.
+-- [Split UTxO]: Add several transaction outputs with addresses belonging to the
+--   wallet. Use the v'Balanced' option to distribute funds evenly.
+-- [Spending particular UTxO]: To ensure that a particular UTxO(s) is/are spent,
+--   add them to the 'TxIn' set of the initial @tx@.
+-- [Get fee]: Caller should calculate the difference of inputs and outputs
+--   in the returned transaction.
```

Comment:
I'm not sure if we should leave the computation or recording of the fee to the caller? 🤔 The `xyzzy` algorithm has to deal with fees anyway, when it applies `marginalCostOfOutput` and (in the unlucky case) attempts to cover the cost with an additional input. In other words, the algorithm (in this form at least), has to deal with fees in a recursive fashion anyway, and it makes sense to take all the responsibility away from the caller. Then again, this would probably involve adding an additional `cost :: tx -> Cost tx` to the `Params` type. 🤔